changefeedccl: Cleanup resources when closing file

miretskiy opened the pull request:
Ensure resources acquired by cloud storage files are released when the sink is closed.

As of https://github.com/cockroachdb/cockroach/pull/88635, cloud storage uses faster implementation of gzip compression algorithm (along with zstd).  This new implementation is sufficiently different from the standard gzip implementation in that it requires the compression codec to be closed, even when the caller is terminating.  Failure to do so results in the memory as well as the goroutine leakage.

This resource leakage may become sufficiently noticable if the changefeed experiences many repeated errors.

This PR modifies Close() call to make sure that the underlying compression codecs are also closed (Note: we rely on the high level logic in distSQL to ensure that the processor gets orderly shut down, and the shutdown code calls Close() method; However, there is still exists a possiblity that the shutdown might not be orderly, and in those cases resource leakage may still occur.  This possiblity will need to be revisited in the follow on PR).

Fixes #106774

Release note (enterprise change): Fix an issue where the changefeeds emitting to cloud sink with compression may experience resource leakage (memory and go routines) when experiencing transient errors.

Commit:
changefeedccl: Cleanup resources when closing file

Ensure resources acquired by cloud storage files are released
when the sink is closed.

As of https://github.com/cockroachdb/cockroach/pull/88635,
cloud storage uses faster implementation of gzip compression
algorithm (along with zstd).  This new implementation
is sufficiently different from the standard gzip implementation
in that it requires the compression codec to be closed, even
when the caller is terminating.  Failure to do so results in the
memory as well as the goroutine leakage.

This resource leakage may become sufficiently noticable
if the changefeed experiences many repeated errors.

This PR modifies Close() call to make sure that the underlying
compression codecs are also closed (Note: we rely on the high level
logic in distSQL to ensure that the processor gets orderly shut down,
and the shutdown code calls Close() method; However, there is still
exists a possiblity that the shutdown might not be orderly, and in those
cases resource leakage may still occur.  This possiblity will need
to be revisited in the follow on PR).

Fixes #106774

Release note (enterprise change): Fix an issue where the changefeeds
emitting to cloud sink with compression may experience resource leakage
(memory and go routines) when experiencing transient errors.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/106795)
<!-- Reviewable:end -->

Commit:
changefeedccl: Release compression codec as soon as error occurs

Proactively close compression codecs if an error occurs
when writing to cloud storage file.

While we always call Close on the sink object, this PR
ensures that even if this doesn't happen, the resources used
by compression code are released as soon as an error (during write)
occurs.

Informs #106774

Release note: None

jayshrivastava reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106795#-:-N_KLwif5A7KNYIZ_GKn:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt and @HonoreDB)



<!-- Sent from Reviewable.io -->

miretskiy commented:
22.1 didn't have this functionality; so ignore backport failure
